Vienna/Sources/Models/Folder.m

```diff
@@ -702,38 +702,45 @@ -(NSArray *)arrayOfUnreadArticlesRefs
                     } else {
                         // some problem
                         NSLog(@"Bug retrieving from cache in folder %li : after %lu insertions of %lu, guid %@",(long)self.itemId, (unsigned long)articles.count,(unsigned long)self.cachedGuids.count,object);
-                        self.isCached = NO;
-                        self.containsBodies = NO;
-                        break;
+                        self.cachedArticles.evictsObjectsWithDiscardedContent = YES;
+                        return [self getArticlesAndCache];
                     }
                 }
                 self.cachedArticles.evictsObjectsWithDiscardedContent = YES;
                 return [articles copy];
             } else {
-                NSArray * articles = [[Database sharedManager] arrayOfArticles:self.itemId filterString:filterString];
-                // Only feeds folders can be cached, as they are the only ones to guarantee
-                // bijection : one article <-> one guid
-                if (self.type == VNAFolderTypeRSS || self.type == VNAFolderTypeOpenReader) {
-                    self.isCached = NO;
-                    self.containsBodies = NO;
-                    [self.cachedArticles removeAllObjects];
-                    [self.cachedGuids removeAllObjects];
-                    for (id object in articles) {
-                        NSString * guid = ((Article *)object).guid;
-                        [self.cachedArticles setObject:object forKey:[NSString stringWithString:guid]];
-                        [self.cachedGuids addObject:guid];
-                    }
-                    self.isCached = YES;
-                    self.containsBodies = YES;
-                }
-                return articles;
-            }
+                return [self getArticlesAndCache];
+           }
         } // synchronized
 	} else {
 	    return [[Database sharedManager] arrayOfArticles:self.itemId filterString:filterString];
     }
 }
 
+/* getArticlesAndCache
+ * Returns articles and store in cache if possible
+ */
+-(NSArray<Article *> *)getArticlesAndCache
```

Comment:
I think this should be split in to two functions:

1. return the array of Articles (if there isn't already a function)
2. cache the articles

The name `getArticlesAndCache` is a bit confusing, and I'd expect two objects returned: 1) array of articles, 2) a cache object

Reply:
The cache object itself is self.cache
It's just a byproduct of the process that gets the content of the articles.
Updated version clarifies this.